<a href="https://colab.research.google.com/github/qianjing2020/DS-Unit-1-Sprint-1-Data-Wrangling-and-Storytelling/blob/master/LS_DS_121_Join_and_Reshape_Data_Assignment_Possible_Solution.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

_Lambda School Data Science_

# Join and Reshape datasets

Objectives
- concatenate data with pandas
- merge data with pandas
-  understand tidy data formatting
-  melt and pivot data with pandas

Links
- [Pandas Cheat Sheet](https://github.com/pandas-dev/pandas/blob/master/doc/cheatsheet/Pandas_Cheat_Sheet.pdf)
- [Tidy Data](https://en.wikipedia.org/wiki/Tidy_data)
  - Combine Data Sets: Standard Joins
  - Tidy Data
  - Reshaping Data
- Python Data Science Handbook
  - [Chapter 3.6](https://jakevdp.github.io/PythonDataScienceHandbook/03.06-concat-and-append.html), Combining Datasets: Concat and Append
  - [Chapter 3.7](https://jakevdp.github.io/PythonDataScienceHandbook/03.07-merge-and-join.html), Combining Datasets: Merge and Join
  - [Chapter 3.8](https://jakevdp.github.io/PythonDataScienceHandbook/03.08-aggregation-and-grouping.html), Aggregation and Grouping
  - [Chapter 3.9](https://jakevdp.github.io/PythonDataScienceHandbook/03.09-pivot-tables.html), Pivot Tables
  
Reference
- Pandas Documentation: [Reshaping and Pivot Tables](https://pandas.pydata.org/pandas-docs/stable/reshaping.html)
- Modern Pandas, Part 5: [Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
!wget https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz

--2019-10-31 21:08:29--  https://s3.amazonaws.com/instacart-datasets/instacart_online_grocery_shopping_2017_05_01.tar.gz
Resolving s3.amazonaws.com (s3.amazonaws.com)... 52.216.12.14
Connecting to s3.amazonaws.com (s3.amazonaws.com)|52.216.12.14|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 205548478 (196M) [application/x-gzip]
Saving to: ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’

instacart_online_gr 100%[===================>] 196.03M  84.3MB/s    in 2.3s    

2019-10-31 21:08:32 (84.3 MB/s) - ‘instacart_online_grocery_shopping_2017_05_01.tar.gz’ saved [205548478/205548478]



In [0]:
!tar --gunzip --extract --verbose --file=instacart_online_grocery_shopping_2017_05_01.tar.gz

instacart_2017_05_01/
instacart_2017_05_01/._aisles.csv
instacart_2017_05_01/aisles.csv
instacart_2017_05_01/._departments.csv
instacart_2017_05_01/departments.csv
instacart_2017_05_01/._order_products__prior.csv
instacart_2017_05_01/order_products__prior.csv
instacart_2017_05_01/._order_products__train.csv
instacart_2017_05_01/order_products__train.csv
instacart_2017_05_01/._orders.csv
instacart_2017_05_01/orders.csv
instacart_2017_05_01/._products.csv
instacart_2017_05_01/products.csv


In [0]:
%cd instacart_2017_05_01

/content/instacart_2017_05_01


In [0]:
!ls -lh *.csv

-rw-r--r-- 1 502 staff 2.6K May  2  2017 aisles.csv
-rw-r--r-- 1 502 staff  270 May  2  2017 departments.csv
-rw-r--r-- 1 502 staff 551M May  2  2017 order_products__prior.csv
-rw-r--r-- 1 502 staff  24M May  2  2017 order_products__train.csv
-rw-r--r-- 1 502 staff 104M May  2  2017 orders.csv
-rw-r--r-- 1 502 staff 2.1M May  2  2017 products.csv


# Assignment

## Join Data Practice

These are the top 10 most frequently ordered products. How many times was each ordered? 

1. Banana
2. Bag of Organic Bananas
3. Organic Strawberries
4. Organic Baby Spinach 
5. Organic Hass Avocado
6. Organic Avocado
7. Large Lemon 
8. Strawberries
9. Limes 
10. Organic Whole Milk

First, write down which columns you need and which dataframes have them.

Next, merge these into a single dataframe.

Then, use pandas functions from the previous lesson to get the counts of the top 10 most frequently ordered products.

## Top 10 Most Frequently Ordered Products
We need product names and counts of how many times they were sold
- product_name is in products.csv
- instances of specific orders are in order_products__prior.csv and order_products__train.csv

In [0]:
import pandas as pd

orders = pd.read_csv('orders.csv')
orders.head()

order_id  user_id  ... order_hour_of_day  days_since_prior_order
0   2539329        1  ...                 8                     NaN
1   2398795        1  ...                 7                    15.0
2    473747        1  ...                12                    21.0
3   2254736        1  ...                 7                    29.0
4    431534        1  ...                15                    28.0

[5 rows x 7 columns]

We actually don't need anything from orders.csv

In [0]:
products = pd.read_csv('products.csv')
products.head()

product_id  ... department_id
0           1  ...            19
1           2  ...            13
2           3  ...             7
3           4  ...             1
4           5  ...            13

[5 rows x 4 columns]

In [0]:
# read in and concatenate the __prior and __train CSVs
order_products__prior = pd.read_csv('order_products__prior.csv')
order_products__train = pd.read_csv('order_products__train.csv')

order_products = pd.concat([order_products__prior, order_products__train])
print(order_products.shape)
order_products.head()

(33819106, 4)


order_id  product_id  add_to_cart_order  reordered
0         2       33120                  1          1
1         2       28985                  2          1
2         2        9327                  3          0
3         2       45918                  4          1
4         2       30035                  5          0

In [0]:
order_products['product_id'].value_counts()

24852    491291
13176    394930
21137    275577
21903    251705
47209    220877
          ...  
42235         1
5255          1
44495         1
36079         1
20759         1
Name: product_id, Length: 49685, dtype: int64

### Try and eyeball duplicate products in a single order

In [0]:
pd.set_option('display.max_rows', 500)
order_products.sort_values(['order_id', 'product_id', 'reordered']).head(100)

order_id  product_id  add_to_cart_order  reordered
2          1       10246                  3          0
1          1       11109                  2          1
5          1       13176                  6          0
7          1       22035                  8          1
4          1       43633                  5          1
6          1       47209                  7          0
0          1       49302                  1          1
3          1       49683                  4          0
7          2        1819                  8          1
2          2        9327                  3          0
5          2       17794                  6          1
1          2       28985                  2          1
4          2       30035                  5          0
0          2       33120                  1          1
6          2       40141                  7          1
8          2       43668                  9          0
3          2       45918                  4          1
15         3       17461                  7          1
13         3       17668                  5          1
11         3       17704                  3          1
12         3       21903                  4          1
10         3       24838                  2          1
16         3       32665                  8          1
9          3       33754                  1          1
14         3       46667                  6          1
21         4       10054                  5          1
26         4       17616                 10          1
22         4       21351                  6          1
23         4       22598                  7          1
27         4       25146                 11          1
18         4       26434                  2          1
20         4       27761                  4          1
28         4       32645                 12          1
24         4       34862                  8          1
19         4       39758                  3          1
25         4       40285                  9          1
29         4       41276                 13          1
17         4       46842                  1          0
41         5        6184                 12          1
39         5        6348                 10          1
55         5        8479                 26          0
37         5        9633                  8          1
45         5       12962                 16          1
30         5       13176                  1          1
36         5       13245                  7          1
31         5       15005                  2          1
48         5       18569                 19          1
43         5       20914                 14          1
34         5       23909                  5          1
47         5       24773                 18          1
38         5       27360                  9          1
33         5       27966                  4          1
44         5       37011                 15          1
53         5       38693                 24          0
40         5       40878                 11          1
49         5       41176                 20          1
46         5       45698                 17          1
52         5       46522                 23          0
51         5       47209                 22          0
32         5       47329                  3          1
42         5       48002                 13          1
50         5       48366                 21          1
35         5       48370                  6          1
54         5       48825                 25          0
57         6       15873                  2          0
56         6       40462                  1          0
58         6       41897                  3          0
59         7       34050                  1          0
60         7       46802                  2          0
61         8       23423                  1          1
74         9         432                 13          1
65         9        2014                  4 

### I'm not seeing any duplicates, but maybe we're just getting unlucky? Can we check for duplicates programmatically?

In [0]:
order_products.duplicated(subset=['order_id', 'product_id']).value_counts()

False    33819106
dtype: int64

In [0]:
# order_products[['order_id', 'product_id']].nunique()

order_id      3346083
product_id      49685
dtype: int64

In [0]:
order_products.shape

(33819106, 4)

In [0]:
order_products.duplicated(subset=['order_id', 'product_id']).value_counts().iloc[0]

33819106

In [0]:
assert order_products.duplicated(subset=['order_id', 'product_id']).value_counts().iloc[0] == order_products.shape[0]

### Conclusion? - This dataset does not have any information about the quantity of items ordered, only unique items ordered and whether the shopper had bought any items in past visits. So our counts of how many times the top 10 products were ordered will really be the number of orders that the top 10 products were included in.

### In order to count the frequency of orders of a given product we need to combine orders and products so that we have names associated with the products in each order.

In [0]:
product_orders = pd.merge(order_products, products[['product_id', 'product_name']])
product_orders.head()

order_id  product_id  add_to_cart_order  reordered        product_name
0         2       33120                  1          1  Organic Egg Whites
1        26       33120                  5          0  Organic Egg Whites
2       120       33120                 13          0  Organic Egg Whites
3       327       33120                  5          1  Organic Egg Whites
4       390       33120                 28          1  Organic Egg Whites

### Getting the final counts is all about just counting up the number of unique instances of product_name

In [0]:
product_orders['product_name'].value_counts().sort_values(ascending=False)[:10]

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

In [0]:
product_orders['product_name'].value_counts().nlargest(10)

Banana                    491291
Bag of Organic Bananas    394930
Organic Strawberries      275577
Organic Baby Spinach      251705
Organic Hass Avocado      220877
Organic Avocado           184224
Large Lemon               160792
Strawberries              149445
Limes                     146660
Organic Whole Milk        142813
Name: product_name, dtype: int64

## Reshape Data Section

- Replicate the lesson code
- Complete the code cells we skipped near the beginning of the notebook
- Table 2 --> Tidy
- Tidy --> Table 2
- Load seaborn's `flights` dataset by running the cell below. Then create a pivot table showing the number of passengers by month and year. Use year for the index and month for the columns. You've done it right if you get 112 passengers for January 1949 and 432 passengers for December 1960.

In [0]:
%matplotlib inline
import pandas as pd
import numpy as np
import seaborn as sns

table1 = pd.DataFrame(
    [[np.nan, 2],
     [16,    11], 
     [3,      1]],
    index=['John Smith', 'Jane Doe', 'Mary Johnson'], 
    columns=['treatmenta', 'treatmentb'])

table2 = table1.T

table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

## Table2 --> Tidy

### Option 1, just transpose it (to get back to the table1 version) and then do the same thing you did to table1

In [0]:
tidy2 = table2.T.reset_index().melt(id_vars='index').rename(columns= {
    'index': 'name', 
    'variable': 'trt', 
    'value': 'result'
})
tidy2

name         trt  result
0    John Smith  treatmenta     NaN
1      Jane Doe  treatmenta    16.0
2  Mary Johnson  treatmenta     3.0
3    John Smith  treatmentb     2.0
4      Jane Doe  treatmentb    11.0
5  Mary Johnson  treatmentb     1.0

In [0]:
tidy2['trt'] = tidy2['trt'].str.replace('treatment', '')
tidy2 = tidy2.set_index('name')
tidy2

trt  result
name                    
John Smith     a     NaN
Jane Doe       a    16.0
Mary Johnson   a     3.0
John Smith     b     2.0
Jane Doe       b    11.0
Mary Johnson   b     1.0

## Tidy --> Table2

Do the same thing you did to table 1, but then transpose it when you're finished

In [0]:
tidy2.pivot_table(index='name', columns='trt', values='result').T

name  Jane Doe  John Smith  Mary Johnson
trt                                     
a         16.0         NaN           3.0
b         11.0         2.0           1.0

### Code from Vera:

In [0]:
table2

John Smith  Jane Doe  Mary Johnson
treatmenta         NaN      16.0           3.0
treatmentb         2.0      11.0           1.0

In [0]:
table3 = table2.reset_index()
table3

index  John Smith  Jane Doe  Mary Johnson
0  treatmenta         NaN      16.0           3.0
1  treatmentb         2.0      11.0           1.0

In [0]:
table4 = table3.melt(id_vars='index')
table4

index      variable  value
0  treatmenta    John Smith    NaN
1  treatmentb    John Smith    2.0
2  treatmenta      Jane Doe   16.0
3  treatmentb      Jane Doe   11.0
4  treatmenta  Mary Johnson    3.0
5  treatmentb  Mary Johnson    1.0

In [0]:
# rename the columns
table5 = table4.rename(columns = {'index': 'treatment', 'variable': 'name',  'value': 'result'})
table5

treatment          name  result
0  treatmenta    John Smith     NaN
1  treatmentb    John Smith     2.0
2  treatmenta      Jane Doe    16.0
3  treatmentb      Jane Doe    11.0
4  treatmenta  Mary Johnson     3.0
5  treatmentb  Mary Johnson     1.0

In [0]:
#reorder the columns
table6 = table5[['name', 'treatment', 'result']]
table6

name   treatment  result
0    John Smith  treatmenta     NaN
1    John Smith  treatmentb     2.0
2      Jane Doe  treatmenta    16.0
3      Jane Doe  treatmentb    11.0
4  Mary Johnson  treatmenta     3.0
5  Mary Johnson  treatmentb     1.0

In [0]:
# replace 'treatment' with ''
table6.treatment = table6.treatment.str.replace('treatment', '')
table6

name treatment  result
0    John Smith         a     NaN
1    John Smith         b     2.0
2      Jane Doe         a    16.0
3      Jane Doe         b    11.0
4  Mary Johnson         a     3.0
5  Mary Johnson         b     1.0

In [0]:
# process back to table 2
table7 = table6.pivot(index='treatment', columns='name', values='result')
table7

name       Jane Doe  John Smith  Mary Johnson
treatment                                    
a              16.0         NaN           3.0
b              11.0         2.0           1.0

In [0]:
flights = sns.load_dataset('flights')
flights.head(15)

year      month  passengers
0   1949    January         112
1   1949   February         118
2   1949      March         132
3   1949      April         129
4   1949        May         121
5   1949       June         135
6   1949       July         148
7   1949     August         148
8   1949  September         136
9   1949    October         119
10  1949   November         104
11  1949   December         118
12  1950    January         115
13  1950   February         126
14  1950      March         141

In [0]:
# flights['passengers'].plot()

In [0]:
wide = flights.pivot(index='year', columns='month', values='passengers')
wide

month  January  February  March  April  ...  September  October  November  December
year                                    ...                                        
1949       112       118    132    129  ...        136      119       104       118
1950       115       126    141    135  ...        158      133       114       140
1951       145       150    178    163  ...        184      162       146       166
1952       171       180    193    181  ...        209      191       172       194
1953       196       196    236    235  ...        237      211       180       201
1954       204       188    235    227  ...        259      229       203       229
1955       242       233    267    269  ...        312      274       237       278
1956       284       277    317    313  ...        355      306       271       306
1957       315       301    356    348  ...        404      347       305       336
1958       340       318    362    348  ...        404      359       310       337
1959       360       342    406    396  ...        463      407       362       405
1960       417       391    419    461  ...        508      461       390       432

[12 rows x 12 columns]

In [0]:
wide = flights.pivot_table(index='year', columns='month', values='passengers')
wide

month  January  February  March  April  ...  September  October  November  December
year                                    ...                                        
1949       112       118    132    129  ...        136      119       104       118
1950       115       126    141    135  ...        158      133       114       140
1951       145       150    178    163  ...        184      162       146       166
1952       171       180    193    181  ...        209      191       172       194
1953       196       196    236    235  ...        237      211       180       201
1954       204       188    235    227  ...        259      229       203       229
1955       242       233    267    269  ...        312      274       237       278
1956       284       277    317    313  ...        355      306       271       306
1957       315       301    356    348  ...        404      347       305       336
1958       340       318    362    348  ...        404      359       310       337
1959       360       342    406    396  ...        463      407       362       405
1960       417       391    419    461  ...        508      461       390       432

[12 rows x 12 columns]

### I given an example of both [pd.pivot](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pivot.html) and [pd.pivot_table](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.DataFrame.pivot_table.html) in this instance the both give the same result but that is not always the case.

### `pivot_table` is the more powerful version of the function as it can handle multiple indexes (called a hierarchical index) and duplicate-value row/column combinations.

### Stack Overflow Question about this: <https://stackoverflow.com/questions/30960338/pandas-difference-between-pivot-and-pivot-table-why-is-only-pivot-table-workin>

## Join Data Stretch Challenge

The [Instacart blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2) has a visualization of "**Popular products** purchased earliest in the day (green) and latest in the day (red)." 

The post says,

> "We can also see the time of day that users purchase specific products.

> Healthier snacks and staples tend to be purchased earlier in the day, whereas ice cream (especially Half Baked and The Tonight Dough) are far more popular when customers are ordering in the evening.

> **In fact, of the top 25 latest ordered products, the first 24 are ice cream! The last one, of course, is a frozen pizza.**"

Your challenge is to reproduce the list of the top 25 latest ordered popular products.

We'll define "popular products" as products with more than 2,900 orders.



In [0]:
##### YOUR CODE HERE #####

## Reshape Data Stretch Challenge

_Try whatever sounds most interesting to you!_

- Replicate more of Instacart's visualization showing "Hour of Day Ordered" vs "Percent of Orders by Product"
- Replicate parts of the other visualization from [Instacart's blog post](https://tech.instacart.com/3-million-instacart-orders-open-sourced-d40d29ead6f2), showing "Number of Purchases" vs "Percent Reorder Purchases"
- Get the most recent order for each user in Instacart's dataset. This is a useful baseline when [predicting a user's next order](https://www.kaggle.com/c/instacart-market-basket-analysis)
- Replicate parts of the blog post linked at the top of this notebook: [Modern Pandas, Part 5: Tidy Data](https://tomaugspurger.github.io/modern-5-tidy.html)

In [0]:
##### YOUR CODE HERE #####